workflow-bot-app/src/main/java/com/symphony/bdk/workflow/monitoring/service/MonitoringService.java

```diff
@@ -100,42 +112,95 @@ public WorkflowActivitiesView listWorkflowInstanceActivities(String workflowId,
+      Map<String, WorkflowNode> dictionary, String key, WorkflowNode workflowNode,
+      WorkflowDirectGraph.NodeChildren children, List<String> parents) {
+    NodeDefinitionView.NodeDefinitionViewBuilder nodeBuilder = NodeDefinitionView.builder();
+    // check if the current node is part of the gateway
+    List<String> realParents = new ArrayList<>();
+    for (String parent : parents) {
+      if (directGraph.getChildren(parent).getChildren().size() > 1) {
+        realParents.add(parent + GATEWAY_SUFFIX);
+      } else {
+        realParents.add(parent);
+      }
+    }
+
+    if (children.isEmpty()) {
+      nodeBuilder.nodeId(key).parents(realParents).children(Collections.emptyList());
+    } else if (children.isChildUnique()) {
+      nodeBuilder.nodeId(key)
+          .parents(realParents)
+          .children(List.of(NodeDefinitionView.ChildView.of(children.getUniqueChild(),
+              determineCondition(dictionary, key, children.getUniqueChild()))));
+    } else {
+      NodeDefinitionView gateway = NodeDefinitionView.builder()
+          .nodeId(key + GATEWAY_SUFFIX)
+          .parents(List.of(key))
+          .group(GATEWAY)
+          .type(children.getGateway() == Gateway.PARALLEL ? GATEWAY_ALL_OF : GATEWAY_ONE_OF)
+          .children(children.getChildren().stream().map(node -> {
+            String condition = determineCondition(dictionary, key, node);
+            return NodeDefinitionView.ChildView.of(node, condition);
+          }).collect(Collectors.toList()))
+          .build();
+      activities.add(gateway);
+      nodeBuilder.nodeId(key)
+          .parents(realParents)
+          .children(List.of(NodeDefinitionView.ChildView.of(gateway.getNodeId())));
+    }
+
+    if (workflowNode.getActivity() != null) {
+      TaskTypeEnum taskTypeEnum = findByAbbr(workflowNode.getActivity().getClass().getSimpleName());
+      nodeBuilder.type(taskTypeEnum.toType());
+      nodeBuilder.group(taskTypeEnum.toGroup());
+    } else if (workflowNode.getEvent() != null) {
+      TaskTypeEnum taskTypeEnum = findByAbbr(workflowNode.getEvent().getEventType());
+      nodeBuilder.type(taskTypeEnum.toType());
+      nodeBuilder.group(taskTypeEnum.toGroup());
+    } else if (workflowNode.getElementType() == WorkflowNodeType.JOIN_ACTIVITY) {
+      nodeBuilder.type(GATEWAY_JOIN);
+      nodeBuilder.group(GATEWAY);
+    }
+    return nodeBuilder.build();
+  }
+
+  @Nullable
+  private static String determineCondition(Map<String, WorkflowNode> dictionary, String key, String node) {
+    WorkflowNode child = dictionary.get(node);
+    String condition = child.getIfCondition(key);
+    if (condition == null) {
+      condition = child.getElementType() == WorkflowNodeType.ACTIVITY_EXPIRED_EVENT ? "expired"
+          : child.getElementType() == WorkflowNodeType.ACTIVITY_FAILED_EVENT ? "failed" : null;
+    }
+    return condition;
+  }
```

Comment:
I suggest to directly pass the child instead of dictionary+node, so we only pass the required objects.
```suggestion
@Nullable
private static String determineCondition(WorkflowNode child, String key) {
    String condition = child.getIfCondition(key);
    if (condition == null) {
      condition = child.getElementType() == WorkflowNodeType.ACTIVITY_EXPIRED_EVENT ? "expired"
          : child.getElementType() == WorkflowNodeType.ACTIVITY_FAILED_EVENT ? "failed" : null;
    }
    return condition;
  }
```

Reply:
> I suggest to directly pass the child instead of dictionary+node, so we only pass the required objects.

sure. agree, looks like Intellji extract code is not that intelligent

Reply:
@symphony-soufiane done as suggested